### PR TITLE
[WIP] Various improvements for DQN and the replay memory

### PR DESCRIPTION
**Pull request by matthiasplappert**
- Fix various problems with sequential memory
- Fix various problems with `CEMAgent`
- Remove duplicate forward pass from DQN agent
- Remove performance bottleneck from `TrainIntervalLogger`
- Add tests for core and memory classes
- Add integration test for CEM

### CODE REVIEW

File: rl/memory.py [hunk 1]
```diff
@@ -45,52 +67,83 @@ def __init__(self, limit):
...
+                terminal0 = self.terminals[idx - 2] if idx >= 2 else False
+            assert 1 <= idx < self.nb_entries
+
             # This code is slightly complicated by the fact that subsequent observations might be
             # from different episodes. We ensure that an experience never spans multiple episodes.
             # This is probably not that important in practice but it seems cleaner.
             state0 = [self.observations[idx - 1]]
-            for current_idx in range(idx - 2, idx - window_length - 1, -1):
-                if self.terminals[current_idx]:
+            for offset in range(0, window_length - 1):
```
Comment (SimonRamstedt) [inline rl/memory.py]: This seems to be the same logic as in get get_recent_state(). Combine into a common function get_state() ?

**Comment by RyanHope** [comment]
My dqn model that works in master is failing on this branch. I will try and do some investigation on this but I wanted to report it first...

```
ryan@Redman:~/workspace/DeepSF$ python dqn.py -d ~/Dropbox/cmu-sf/deepsf-data2/ --visualize -r rawpnts
Using Theano backend.
Using gpu device 0: GeForce GTX 1060 3GB (CNMeM is enabled with initial size: 10.0% of memory, cuDNN not available)
____________________________________________________________________________________________________
Layer (type)                     Output Shape          Param #     Connected to                     
====================================================================================================
flatten_1 (Flatten)              (None, 60)            0           flatten_input_1[0][0]            
____________________________________________________________________________________________________
dense_1 (Dense)                  (None, 64)            3904        flatten_1[0][0]                  
____________________________________________________________________________________________________
elu_1 (ELU)                      (None, 64)            0           dense_1[0][0]                    
____________________________________________________________________________________________________
dense_2 (Dense)                  (None, 64)            4160        elu_1[0][0]                      
____________________________________________________________________________________________________
elu_2 (ELU)                      (None, 64)            0           dense_2[0][0]                    
____________________________________________________________________________________________________
dense_3 (Dense)                  (None, 64)            4160        elu_2[0][0]                      
____________________________________________________________________________________________________
elu_3 (ELU)                      (None, 64)            0           dense_3[0][0]                    
____________________________________________________________________________________________________
dense_4 (Dense)                  (None, 3)             195         elu_3[0][0]                      
____________________________________________________________________________________________________
activation_1 (Activation)        (None, 3)             0           dense_4[0][0]                    
====================================================================================================
Total params: 12419
____________________________________________________________________________________________________
None
Training for 27000000 steps ...
     2700/27000000: episode: 1, duration: 5.134s, episode steps: 2700, steps per second: 526, episode reward: -1282.000, mean reward: -0.475 [-100.000, 100.000], mean action: 1.164 [0.000, 2.000], mean observation: 72.092 [-304.000, 549.654], loss: --, mean_absolute_error: --, mean_q_new: --
Traceback (most recent call last):
  File "dqn.py", line 97, in <module>
    main(args)
  File "dqn.py", line 78, in main
    agent.fit(env, nb_steps=STEPS_PER_EPISODE*10000, visualize=True, verbose=2, action_repetition=args.frameskip, callbacks=[log])
  File "/home/ryan/workspace/keras-rl/rl/core.py", line 97, in fit
    metrics = self.backward(reward, terminal=done)
  File "/home/ryan/workspace/keras-rl/rl/agents/dqn.py", line 279, in backward
    metrics = self.trainable_model.train_on_batch([state0_batch, targets, masks], dummy_targets)
  File "/usr/local/lib/python2.7/dist-packages/Keras-1.1.0-py2.7.egg/keras/engine/training.py", line 1221, in train_on_batch
    outputs = self.train_function(ins)
  File "/usr/local/lib/python2.7/dist-packages/Keras-1.1.0-py2.7.egg/keras/backend/theano_backend.py", line 717, in __call__
    return self.function(*inputs)
  File "/home/ryan/.local/lib/python2.7/site-packages/theano/compile/function_module.py", line 871, in __call__
    storage_map=getattr(self.fn, 'storage_map', None))
  File "/home/ryan/.local/lib/python2.7/site-packages/theano/gof/link.py", line 314, in raise_with_op
    reraise(exc_type, exc_value, exc_trace)
  File "/home/ryan/.local/lib/python2.7/site-packages/theano/compile/function_module.py", line 859, in __call__
    outputs = self.fn()
ValueError: GpuElemwise. Input dimension mis-match. Input 1 (indices start at 0) has shape[1] == 1, but the output's size on that axis is 32.
Apply node that caused the error: GpuElemwise{Composite{Abs((i0 - i1))},no_inplace}(GpuDimShuffle{x,0}.0, GpuFromHost.0)
Toposort index: 92
Inputs types: [CudaNdarrayType(float32, row), CudaNdarrayType(float32, matrix)]
Inputs shapes: [(1, 32), (32, 1)]
Inputs strides: [(0, 1), (1, 0)]
Inputs values: ['not shown', 'not shown']
Outputs clients: [[GpuCAReduce{add}{1,1}(GpuElemwise{Composite{Abs((i0 - i1))},no_inplace}.0)]]

HINT: Re-running with most Theano optimization disabled could give you a back-trace of when this node was created. This can be done with by setting the Theano flag 'optimizer=fast_compile'. If that does not work, Theano optimizations can be disabled with 'optimizer=None'.
HINT: Use the Theano flag 'exception_verbosity=high' for a debugprint and storage map footprint of this apply node.
```

**Comment by matthiasplappert** [comment]
Yes, there's currently a bug if you use the Theano backend and with a metric. This will be fixed before this gets merged.

**Comment by RyanHope** [comment]
I am pretty sure I tracked it down to this commit -> 8231b72ead028e7ff1e1c7c51bfa8c7587151711

**Comment by matthiasplappert** [comment]
Yes this is because of the way we mask the targets in the Lamba layer. I have a plan how this could be addressed siding worry about it too much for now.

**Comment by RyanHope** [comment]
Ok I just figured I would help test what seems like the dev branch because when I start submitting PRs for new agents and feature requests I wanted to have a stable dev branch to work off of.

**Pull request merged**